### PR TITLE
remove config.EVMNetworkID()

### LINKED ISSUE #3452: [blockdao] remove config.EVMNetworkID()
### Description of the code refactor
solve the 2 TODO in commit https://github.com/iotexproject/iotex-core/commit/4be5188a986fcf74d88f6a8a5ce8cd7e31760517:
1. https://github.com/iotexproject/iotex-core/commit/4be5188a986fcf74d88f6a8a5ce8cd7e31760517#diff-f082988255fcbe6608ac6f4488621d8d47ea8c032ca7235652a79c895a618636R45
2. https://github.com/iotexproject/iotex-core/commit/4be5188a986fcf74d88f6a8a5ce8cd7e31760517#diff-3f4c4a2bbdc33fe76d0036bcaefaca49d48ddab45a3406eab8cfd0761da4b47fR266

pass the EVM network ID (from config) at time of newFileDAO(), feel free to discuss with me for the background and context of this task

### Why is this needed
- [] better/less memory usage
- [] reduce running time
- [x] clean (easy-to-understand) data structure or algorithm
- [] remove dead code
- [x] better code design or structure (like reduce complexity)
- [x] reduce/decouple dependencies between code modules
- [] increase test coverage
- [] better code readability

### PR DESCRIPTION
**Pull request by millken**
# Description

Fixes #3452

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: blockchain/block/blockstore.go [hunk 1]
```diff
@@ -41,9 +40,8 @@ func (in *Store) ToProto() *iotextypes.BlockStore {
 }
 
 // FromProto converts from proto message
-func (in *Store) FromProto(pb *iotextypes.BlockStore) error {
-	// TODO: pass the correct EVM network ID at the time of newFileDAOv2()
-	blk, err := (&Deserializer{}).SetEvmNetworkID(config.EVMNetworkID()).FromBlockProto(pb.Block)
+func (in *Store) FromProto(evmNetworkID uint32, pb *iotextypes.BlockStore) error {
```
Comment (CoderZhi) [inline blockchain/block/blockstore.go:43]: if a store only contains the data of the same `evmNetworkID`, then the value should be stored in `Store`.

File: blockchain/block/blockstore.go [hunk 2]
```diff
@@ -41,9 +40,8 @@ func (in *Store) ToProto() *iotextypes.BlockStore {
...
+	blk, err := (&Deserializer{}).SetEvmNetworkID(evmNetworkID).FromBlockProto(pb.Block)
```
Comment (CoderZhi) [inline blockchain/block/blockstore.go:44]: moreover, I don't see the value of creating a new `Deserializer` in each function call

File: blockchain/filedao/filedao_v2.go [hunk 3]
```diff
@@ -48,11 +48,12 @@ type (
 		hashStore db.CountingIndex // store block hash
 		blkStore  db.CountingIndex // store raw blocks
 		sysStore  db.CountingIndex // store transaction log
+		cfg       ModuleConfig
 	}
 )
 
 // newFileDAOv2 creates a new v2 file
-func newFileDAOv2(bottom uint64, cfg db.Config) (*fileDAOv2, error) {
+func newFileDAOv2(bottom uint64, cfg ModuleConfig) (*fileDAOv2, error) {
```
Comment (Liuhaai) [inline blockchain/filedao/filedao_v2.go:56]: Get the deserializer from chainbuilder when a new filedao is created .

**Review by Liuhaai — CHANGES_REQUESTED** [review]

File: blockchain/filedao/filedao_v2.go (hunk 3 above)
Comment (Liuhaai) [reply blockchain/filedao/filedao_v2.go:56]: we can have an offline discussion if you have a better solution here

File: blockchain/filedao/staging_buffer.go [hunk 4]
```diff
@@ -40,7 +40,7 @@ func (s *stagingBuffer) Put(pos uint64, blkBytes []byte) (bool, error) {
 		return false, ErrNotSupported
 	}
 	blk := &block.Store{}
-	if err := blk.Deserialize(blkBytes); err != nil {
+	if err := blk.Deserialize(0, blkBytes); err != nil {
```
[the remark this answers is not in the document]
Comment (millken) [reply blockchain/filedao/staging_buffer.go]: Updated

File: blockchain/block/blockstore.go [hunk 5]
```diff
@@ -12,17 +12,26 @@ import (
...
+		EVMNetworkID uint32
+		Block        *Block
+		Receipts     []*action.Receipt
 	}
 )
 
+// NewStore creates a new Store
+func NewStore(evmNetworkID uint32) *Store {
+	return &Store{
+		EVMNetworkID: evmNetworkID,
```
Comment (CoderZhi) [inline blockchain/block/blockstore.go:29]: I am sorry of my last comment. I was mislead by the *BAD* naming of this struct. 

How could it be named as `store`?!

**Review by CoderZhi — COMMENTED** [review]
@millken My last comment about `store` is wrong. Sorry about that.

File: blockchain/block/blockstore.go (hunk 5 above)
Comment (millken) [reply blockchain/block/blockstore.go:29]: You mean it should be renamed to a more understandable name?

File: blockchain/blockdao/blockdao.go [hunk 6]
```diff
@@ -57,11 +57,51 @@ type (
 		footerCache  *cache.ThreadSafeLruCache
 		tipHeight    uint64
 	}
+	// ModuleConfig represents the configuration of File DAO
+	ModuleConfig struct {
+		db.Config
+		Option *configOption
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:63]: evmNetworkID uint32

File: blockchain/blockdao/blockdao.go [hunk 7]
```diff
@@ -57,11 +57,51 @@ type (
...
+	}
+
+	// ModuleConfigOption applies the configuration of File DAO
+	ModuleConfigOption func(*configOption) error
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:67]: func(*ModuleConfig) error

File: blockchain/blockdao/blockdao.go [hunk 8]
```diff
@@ -57,11 +57,51 @@ type (
...
+
+	configOption struct {
+		EvmNetworkID uint32
+	}
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:71]: no need this struct

File: blockchain/blockdao/blockdao.go [hunk 9]
```diff
@@ -57,11 +57,51 @@ type (
 		footerCache  *cache.ThreadSafeLruCache
 		tipHeight    uint64
 	}
+	// ModuleConfig represents the configuration of File DAO
+	ModuleConfig struct {
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:61]: maybe name it `BlockdaoConfig`, in the future we'll have similar config for other modules

File: blockchain/blockdao/blockdao.go [hunk 10]
```diff
@@ -57,11 +57,51 @@ type (
...
+// EVMNetworkIDOption sets the EVM network ID
+func EVMNetworkIDOption(evmNetworkID uint32) ModuleConfigOption {
+	return func(opt *configOption) error {
+		opt.EvmNetworkID = evmNetworkID
+		return nil
+	}
+}
+
+// CreateModuleConfig creates a ModuleConfig
+func CreateModuleConfig(cfg db.Config, options ...ModuleConfigOption) (ModuleConfig, error) {
```
Comment (dustinxie) [inline blockchain/blockdao/blockdao.go:83]: CreateBlockdaoConfig

File: blockchain/filedao/filedao.go [hunk 11]
```diff
@@ -75,15 +75,51 @@ type (
...
 		splitHeight uint64
-		cfg         db.Config
+		cfg         ModuleConfig
 		currFd      BaseFileDAO
 		legacyFd    FileDAO
 		v2Fd        *FileV2Manager // a collection of v2 db files
 	}
+
+	// ModuleConfig represents the configuration of File DAO
+	ModuleConfig struct {
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:85]: FiledaoConfig

File: blockchain/filedao/filedao.go [hunk 12]
```diff
@@ -75,15 +75,51 @@ type (
...
+		db.Config
+		Option *configOption
+	}
+
+	// ModuleConfigOption applies the configuration of File DAO
+	ModuleConfigOption func(*configOption) error
+
+	configOption struct {
+		evmNetworkID uint32
+	}
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:95]: same as above

File: blockchain/filedao/filedao.go [hunk 13]
```diff
@@ -75,15 +75,51 @@ type (
...
+// EVMNetworkIDOption sets the EVM network ID
+func EVMNetworkIDOption(evmNetworkID uint32) ModuleConfigOption {
+	return func(opt *configOption) error {
+		opt.evmNetworkID = evmNetworkID
+		return nil
+	}
+}
+
+// CreateModuleConfig creates a ModuleConfig
+func CreateModuleConfig(cfg db.Config, options ...ModuleConfigOption) (ModuleConfig, error) {
```
Comment (dustinxie) [inline blockchain/filedao/filedao.go:107]: CreateFiledaoConfig

File: blockchain/filedao/filedao_v2_util.go [hunk 14]
```diff
@@ -34,7 +34,7 @@ func (fd *fileDAOv2) populateStagingBuffer() (*stagingBuffer, error) {
 		if err != nil {
 			return nil, err
 		}
-		info := &block.Store{}
+		info := block.NewStore(fd.cfg.Option.evmNetworkID)
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:37]: add a common deserializer into `fileDAOv2`, and here:
```
info, err := fd.deser.DeserializeBlockStore(v)
```

File: blockchain/filedao/filedao_v2_util.go [hunk 15]
```diff
@@ -206,11 +206,11 @@ func (fd *fileDAOv2) getBlockStore(height uint64) (*block.Store, error) {
...
 	// add to read cache
 	fd.blkCache.Add(storeKey, pbStores)
-	return extractBlockStore(pbStores, stagingKey(height, fd.header))
+	return extractBlockStore(fd.cfg.Option.evmNetworkID, pbStores, stagingKey(height, fd.header))
 }
 
-func extractBlockStore(pbStores *iotextypes.BlockStores, height uint64) (*block.Store, error) {
-	info := &block.Store{}
+func extractBlockStore(evmNetworkID uint32, pbStores *iotextypes.BlockStores, height uint64) (*block.Store, error) {
+	info := block.NewStore(evmNetworkID)
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:213]: add a common deserializer into `fileDAOv2`, and here:
```
info, err := fd.deser.FromBlockStoreProto(v)
```

File: blockchain/block/blockstore.go [hunk 16]
```diff
@@ -12,17 +12,26 @@ import (
...
 	"github.com/iotexproject/iotex-core/action"
-	"github.com/iotexproject/iotex-core/config"
 )
 
 type (
 	// Store defines block storage schema
 	Store struct {
-		Block    *Block
-		Receipts []*action.Receipt
+		EVMNetworkID uint32
```
Comment (dustinxie) [inline blockchain/block/blockstore.go:20]: not needed, see comment in `blockchain/filedao/filedao_v2_util.go`

File: blockchain/block/blockstore.go (hunk 5 above)
Comment (CoderZhi) [reply blockchain/block/blockstore.go:29]: let's leave the `Store` as it is. Do not change anything else in `blockdao` and `block`. Let's focus on deleting `config.EVMNetworkID` in this PR.
So, ignore my previous comments, your implementation is fine.

File: chainservice/builder.go [hunk 17]
```diff
@@ -251,8 +251,8 @@ func (builder *Builder) buildBlockDAO(forTest bool) error {
 		dbConfig := builder.cfg.DB
 		dbConfig.DbPath = builder.cfg.Chain.ChainDBPath
 		dbConfig.CompressLegacy = builder.cfg.Chain.CompressBlock
-
-		builder.cs.blockdao = blockdao.NewBlockDAO(indexers, dbConfig)
+		blockDAOConfig, _ := blockdao.CreateBlockdaoConfig(dbConfig, blockdao.EVMNetworkIDOption(builder.cfg.Chain.EVMNetworkID))
```
Comment (CoderZhi) [inline chainservice/builder.go:254]: In product, `EVMNetworkID` will always be assigned to initialize `blockdao`. Therefore, it is not an option, but a parameter. Accordingly, we may define
`blockdao.NewBlockDAO(indexers, evmNetworkID, dbConfig)` and the data structs `XXXConfig` are useless.

Comment (millken) [reply chainservice/builder.go]: done

File: blockchain/filedao/filedao.go [hunk 18]
```diff
@@ -75,15 +75,46 @@ type (
...
+		if err := opt(&cf); err != nil {
+			return cf, err
+		}
+	}
+	return cf, nil
+}
+
 // NewFileDAO creates an instance of FileDAO
-func NewFileDAO(cfg db.Config) (FileDAO, error) {
+func NewFileDAO(cfg Config) (FileDAO, error) {
```
Comment (CoderZhi) [inline blockchain/filedao/filedao.go:117]: similar to blockdao, filedao's config may not be essential.

File: blockchain/filedao/filedao_legacy.go [hunk 19]
```diff
@@ -51,7 +50,7 @@ type (
 	fileDAOLegacy struct {
 		compressBlock bool
 		lifecycle     lifecycle.Lifecycle
-		cfg           db.Config
+		cfg           Config
```
Comment (CoderZhi) [inline blockchain/filedao/filedao_legacy.go:53]: same here

File: blockchain/filedao/filedao_v2.go [hunk 20]
```diff
@@ -48,11 +48,13 @@ type (
...
 		blkStore  db.CountingIndex // store raw blocks
 		sysStore  db.CountingIndex // store transaction log
+		cfg       Config
+		deser     *block.Deserializer
 	}
 )
 
 // newFileDAOv2 creates a new v2 file
-func newFileDAOv2(bottom uint64, cfg db.Config) (*fileDAOv2, error) {
+func newFileDAOv2(bottom uint64, cfg Config) (*fileDAOv2, error) {
```
Comment (CoderZhi) [inline blockchain/filedao/filedao_v2.go:57]: and here

File: blockchain/block/blockstore.go [hunk 21]
```diff
@@ -42,7 +42,6 @@ func (in *Store) ToProto() *iotextypes.BlockStore {
 
 // FromProto converts from proto message
 func (in *Store) FromProto(pb *iotextypes.BlockStore) error {
-	// TODO: pass the correct EVM network ID at the time of newFileDAOv2()
```
Comment (CoderZhi) [inline blockchain/block/blockstore.go:45]: revert

File: blockchain/block/block_deserializer.go [hunk 22]
```diff
@@ -92,3 +92,35 @@ func (bd *Deserializer) DeserializeBody(buf []byte) (*Body, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	// verify merkle root can match after deserialize
+	if err := blk.VerifyTxRoot(); err != nil {
+		return nil, err
+	}
+
+	in.Block = blk
+	in.Receipts = []*action.Receipt{}
```
Comment (dustinxie) [inline blockchain/block/block_deserializer.go:109]: this line not needed, `in.Receipts` is nil and will be allocated in the `append()` below

File: blockchain/block/block_deserializer.go [hunk 23]
```diff
@@ -92,3 +92,35 @@ func (bd *Deserializer) DeserializeBody(buf []byte) (*Body, error) {
...
+	for _, receiptPb := range pb.Receipts {
+		receipt := &action.Receipt{}
+		receipt.ConvertFromReceiptPb(receiptPb)
+		in.Receipts = append(in.Receipts, receipt)
+	}
+
```
Comment (dustinxie) [inline blockchain/block/block_deserializer.go:115]: remove empty line

File: blockchain/block/blockstore.go [hunk 24]
```diff
@@ -42,7 +42,6 @@ func (in *Store) ToProto() *iotextypes.BlockStore {
 
 // FromProto converts from proto message
 func (in *Store) FromProto(pb *iotextypes.BlockStore) error {
```
Comment (CoderZhi) [inline blockchain/block/blockstore.go:44]: are this function and the deserialize function still in use? if not, we can delete them

**Review by CoderZhi — COMMENTED** [review]
the others look fine

File: blockchain/block/blockstore.go (hunk 24 above)
Comment (dustinxie) [inline blockchain/block/blockstore.go:44]: still need to make it `FromProto(pb *iotextypes.BlockStore, evmid uint32)` so we can remove the `config.EVMNetworkID()` in line 45

File: blockchain/filedao/filedao_legacy.go [hunk 25]
```diff
@@ -61,11 +60,11 @@ type (
 )
```
Comment (dustinxie) [inline blockchain/filedao/filedao_legacy.go:60]: similar to `filedao_v2`, add `deser *Deserializer` here

File: blockchain/filedao/filedao_legacy.go [hunk 26]
```diff
@@ -61,11 +60,11 @@ type (
...
 
 // newFileDAOLegacy creates a new legacy file
-func newFileDAOLegacy(cfg db.Config) (FileDAO, error) {
+func newFileDAOLegacy(cfg Config) (FileDAO, error) {
 	return &fileDAOLegacy{
 		compressBlock: cfg.CompressLegacy,
 		cfg:           cfg,
-		kvStore:       db.NewBoltDB(cfg),
+		kvStore:       db.NewBoltDB(cfg.Config),
 		kvStores:      cache.NewThreadSafeLruCache(0),
```
Comment (dustinxie) [inline blockchain/filedao/filedao_legacy.go:68]: and initialize `deser` here

File: blockchain/filedao/filedao_legacy.go [hunk 27]
```diff
@@ -263,8 +262,7 @@ func (fd *fileDAOLegacy) body(h hash.Hash256) (*block.Body, error) {
 		// block body could be empty
 		return &block.Body{}, nil
 	}
-	// TODO: pass the correct EVM network ID at the time of newFileDAOLegacy()
-	return (&block.Deserializer{}).SetEvmNetworkID(config.EVMNetworkID()).DeserializeBody(value)
+	return (&block.Deserializer{}).SetEvmNetworkID(fd.cfg.evmNetworkID).DeserializeBody(value)
```
Comment (dustinxie) [inline blockchain/filedao/filedao_legacy.go:265]: so here just `fd.deser.DeserializeBody()`

File: blockchain/filedao/filedao_v2.go [hunk 28]
```diff
@@ -48,11 +48,13 @@ type (
 		hashStore db.CountingIndex // store block hash
 		blkStore  db.CountingIndex // store raw blocks
 		sysStore  db.CountingIndex // store transaction log
+		cfg       Config
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2.go:51]: we don't need to add `cfg` into the struct? the only new thing needed is `deser *block.Deserializer` below

File: blockchain/filedao/staging_buffer.go [hunk 29]
```diff
@@ -16,15 +16,17 @@ import (
 
 type (
 	stagingBuffer struct {
-		size   uint64
-		buffer []*block.Store
+		evmNetworkID uint32
+		size         uint64
+		buffer       []*block.Store
```
Comment (dustinxie) [inline blockchain/filedao/staging_buffer.go:21]: similar to `filedao_v2`, add `deser *block.Deserializer`

File: blockchain/filedao/staging_buffer.go [hunk 30]
```diff
@@ -16,15 +16,17 @@ import (
...
 )
 
-func newStagingBuffer(size uint64) *stagingBuffer {
+func newStagingBuffer(size uint64, evmNetworkID uint32) *stagingBuffer {
 	return &stagingBuffer{
-		size:   size,
-		buffer: make([]*block.Store, size),
+		evmNetworkID: evmNetworkID,
+		size:         size,
+		buffer:       make([]*block.Store, size),
```
Comment (dustinxie) [inline blockchain/filedao/staging_buffer.go:29]: and initialize `deser` here

File: blockchain/filedao/staging_buffer.go [hunk 31]
```diff
@@ -39,8 +41,9 @@ func (s *stagingBuffer) Put(pos uint64, blkBytes []byte) (bool, error) {
 	if pos >= s.size {
 		return false, ErrNotSupported
 	}
-	blk := &block.Store{}
-	if err := blk.Deserialize(blkBytes); err != nil {
+	deser := (&block.Deserializer{}).SetEvmNetworkID(s.evmNetworkID)
+	blk, err := deser.DeserializeBlockStore(blkBytes)
```
Comment (dustinxie) [inline blockchain/filedao/staging_buffer.go:45]: and `s.deser.DeserializeBlockStore()` here

File: blockchain/filedao/filedao_v2_util.go [hunk 32]
```diff
@@ -206,13 +206,9 @@ func (fd *fileDAOv2) getBlockStore(height uint64) (*block.Store, error) {
...
 }
 
-func extractBlockStore(pbStores *iotextypes.BlockStores, height uint64) (*block.Store, error) {
-	info := &block.Store{}
-	if err := info.FromProto(pbStores.BlockStores[height]); err != nil {
-		return nil, err
-	}
-	return info, nil
+func extractBlockStore(deser *block.Deserializer, pbStores *iotextypes.BlockStores, height uint64) (*block.Store, error) {
+	return deser.FromBlockStoreProto(pbStores.BlockStores[height])
```
Comment (dustinxie) [inline blockchain/filedao/filedao_v2_util.go:213]: can remove this func, b/c it is just 1 line code

File: blockchain/block/blockstore.go (hunk 24 above)
Comment (dustinxie) [reply blockchain/block/blockstore.go:44]: or like @CoderZhi  commented, remove it (and fix test)

File: blockchain/filedao/filedao_v2_util.go (hunk 32 above)
Comment (millken) [reply blockchain/filedao/filedao_v2_util.go]: done

File: blockchain/block/blockstore.go (hunk 24 above)
Comment (millken) [reply blockchain/block/blockstore.go]: done

File: blockchain/filedao/filedao.go [hunk 33]
```diff
@@ -72,26 +72,28 @@ type (
...
-		legacyFd    FileDAO
-		v2Fd        *FileV2Manager // a collection of v2 db files
+		lock         sync.Mutex
+		topIndex     uint64
+		splitHeight  uint64
+		cfg          db.Config
+		currFd       BaseFileDAO
+		legacyFd     FileDAO
+		v2Fd         *FileV2Manager // a collection of v2 db files
+		deser        *block.Deserializer
```
[the remark this answers is not in the document]
Comment (dustinxie) [reply blockchain/filedao/filedao.go:82]: right, can remove

**Comment by CoderZhi** [comment]
fix unit test

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**